Python/optimizer_analysis.c

```diff
@@ -304,11 +777,33 @@ _Py_uop_analyze_and_optimize(
     _PyBloomFilter *dependencies
 )
 {
+    OPT_STAT_INC(optimizer_attempts);
+
     int err = remove_globals(frame, buffer, buffer_size, dependencies);
     if (err <= 0) {
-        return err;
+        goto error;
     }
+
     peephole_opt(frame, buffer, buffer_size);
+
+    err = uop_redundancy_eliminator(
+        (PyCodeObject *)frame->f_executable, buffer,
+        buffer_size, curr_stacklen);
+
+    if (err < 0) {
+        goto error;
+    }
+
+
+
     remove_unneeded_uops(buffer, buffer_size);
+
+    OPT_STAT_INC(optimizer_successes);
     return 1;
+error:
+
+    // The only valid error we can raise is MemoryError.
+    // Other times it's not really errors but things like not being able
+    // to fetch a function version because the function got deleted.
+    return PyErr_Occurred() ? -1 : 0;
```

Comment:
Why do you need to check `PyErr_Occurred()`? `uop_redundancy_eliminator` should not be returning -1 unless an error has occurred.

Reply:
In the old code, constant propagation could raise MemoryError.